### PR TITLE
[DCP] Fixes the stateless optimizer issue of distributed state_dict

### PR DESCRIPTION
**Pull request by fegin**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #135535

Some optimizers don't have states that can cause get_state_dict/set_state_dict behave incorrectly. This PR fixes the issues.

fixes: https://github.com/pytorch/pytorch/issues/133415

cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @LucasLLC @MeetVadakkanchery @mhorowitz @pradeepfn

### CODE REVIEW

File: test/distributed/checkpoint/test_state_dict.py [hunk 1]
```diff
@@ -301,7 +306,7 @@ def test_ddp(self) -> None:
         self.run_subtests(
             {
                 "use_composable": [True, False],
-                "optimizer_class": [torch.optim.Adam, torch.optim.AdamW],
+                "optimizer_class": [torch.optim.SGD],
```
Comment (wz337) [inline test/distributed/checkpoint/test_state_dict.py:309]: Do we want to add torch.optim.SGD to the original list instead?

**Review by wz337 — APPROVED** [review]
LGTM. stamp to unblock

File: test/distributed/checkpoint/test_state_dict.py (hunk 1 above)
Comment (fegin) [reply test/distributed/checkpoint/test_state_dict.py]: good catch

**Pull request closed without merging**